### PR TITLE
Handle ordering of first last aggregation inside aggregator

### PR DESCRIPTION
**Pull request by mustafasrepo**
## Which issue does this PR close?

Closes #.
Improves situation on [#8662](https://github.com/apache/arrow-datafusion/pull/8662)
Related to https://github.com/apache/arrow-datafusion/issues/8582

## Rationale for this change

This PR implements the observation by @alamb at the [PR](https://github.com/apache/arrow-datafusion/pull/8558) that for first and last value aggregation we do not need to sort entire data at its input.  

In other words, This PR is the `FIRST_VALUE` and `LAST_VALUE` aggregation support of the approach 3 in the design [document](https://docs.google.com/document/d/1cIIJL6RKXKge-t8z4Rs_F-XH82giWQfmciwj-h3tho4/edit?usp=sharing)

## Are these changes tested?
Yes

### CODE REVIEW

File: datafusion/sqllogictest/test_files/distinct_on.slt [hunk 1]
```diff
@@ -78,7 +78,7 @@ c 4
 query I
 SELECT DISTINCT ON (c1) c2 FROM aggregate_test_100 ORDER BY c1, c3;
 ----
-5
+4
```
Comment (mustafasrepo) [inline datafusion/sqllogictest/test_files/distinct_on.slt:81]: I ran same query in the postgre, it gave the same result with the new version.

File: datafusion/physical-plan/src/aggregates/mod.rs [hunk 2]
```diff
@@ -1014,17 +942,15 @@ fn aggregate_expressions(
...
-                        let ordering_exprs = ordering_req
-                            .iter()
-                            .map(|item| item.expr.clone())
-                            .collect::<Vec<_>>();
-                        result.extend(ordering_exprs);
-                    }
+                // Append ordering requirements to expressions' results.
+                // This way order sensitive aggregators can satisfy requirement
+                // themselves.
+                if let Some(ordering_req) = agg.order_bys() {
```
Comment (mustafasrepo) [inline datafusion/physical-plan/src/aggregates/mod.rs:948]: Since aggregators themselves handle ordering. We append ordering expression values to the field also for all modes.

File: datafusion/physical-expr/src/aggregate/first_last.rs [hunk 3]
```diff
@@ -459,10 +493,52 @@ impl LastValueAccumulator {
...
+            .zip(self.ordering_req.iter())
+            .map(|(values, req)| {
+                // Take reverse ordering requirement this would enable us to use fetch=1 for last value.
+                SortColumn {
+                    values: values.clone(),
+                    options: Some(!req.options),
+                }
+            })
+            .collect::<Vec<_>>();
+        let indices = lexsort_to_indices(&sort_columns, Some(1))?;
```
Comment (mustafasrepo) [inline datafusion/physical-expr/src/aggregate/first_last.rs:537]: If there is min max alternative to this we can use that one also. However, as far as I know there is no util for this support

File: datafusion/sqllogictest/test_files/groupby.slt [hunk 4]
```diff
@@ -2019,17 +2019,16 @@ SortPreservingMergeExec: [col0@0 ASC NULLS LAST]
...
-----------------ProjectionExec: expr=[col0@2 as col0, col1@3 as col1, col2@4 as col2, col0@0 as col0, col1@1 as col1]
-------------------CoalesceBatchesExec: target_batch_size=8192
---------------------HashJoinExec: mode=Partitioned, join_type=Inner, on=[(col0@0, col0@0)]
-----------------------CoalesceBatchesExec: target_batch_size=8192
-------------------------RepartitionExec: partitioning=Hash([col0@0], 4), input_partitions=1
---------------------------MemoryExec: partitions=1, partition_sizes=[3]
-----------------------CoalesceBatchesExec: target_batch_size=8192
-------------------------RepartitionExec: partitioning=Hash([col0@0], 4), input_partitions=1
---------------------------MemoryExec: partitions=1, partition_sizes=[3]
+------------AggregateExec: mode=Partial, gby=[col0@0 as col0, col1@1 as col1, col2@2 as col2], aggr=[LAST_VALUE(r.col1)]
```
Comment (mustafasrepo) [inline datafusion/sqllogictest/test_files/groupby.slt:2022]: Since first_value and last_value no longer requires ordering at its input. SortExecs are removed from the plan.

File: datafusion/sqllogictest/test_files/groupby.slt [hunk 5]
```diff
@@ -2209,7 +2208,7 @@ ProjectionExec: expr=[a@0 as a, b@1 as b, LAST_VALUE(annotated_data_infinite2.c)
 ----StreamingTableExec: partition_sizes=1, projection=[a, b, c], infinite_source=true, output_ordering=[a@0 ASC NULLS LAST, b@1 ASC NULLS LAST, c@2 ASC NULLS LAST]
 
 query III
-SELECT a, b, LAST_VALUE(c ORDER BY a DESC) as last_c
+SELECT a, b, LAST_VALUE(c ORDER BY a DESC, c ASC) as last_c
```
Comment (mustafasrepo) [inline datafusion/sqllogictest/test_files/groupby.slt:2211]: The result of this test was not unique according to specifications. I changed test to make result unique.

File: datafusion/physical-expr/src/aggregate/first_last.rs (hunk 3 above)
Comment (tustvold) [reply datafusion/physical-expr/src/aggregate/first_last.rs:539]: I'm not aware of a min/max kernel that returns the ordinal position of the min/max

File: datafusion/physical-expr/src/aggregate/first_last.rs [hunk 6]
```diff
@@ -459,10 +493,52 @@ impl LastValueAccumulator {
...
+    fn update_with_new_row(&mut self, row: &[ScalarValue]) -> Result<()> {
+        let value = &row[0];
+        let orderings = &row[1..];
+        // Update when
+        // - no value in the state
+        // - There is no specific requirement, but a new value (most recent entry in terms of execution)
+        // - There is a more recent entry in terms of requirement
+        if !self.is_set
+            || self.orderings.is_empty()
+            || compare_rows(
```
Comment (tustvold) [inline datafusion/physical-expr/src/aggregate/first_last.rs:506]: I'm sure you are aware but https://docs.rs/arrow-row/latest/arrow_row/ will be a much faster way to perform row-based comparisons than relying on ScalarValue

Comment (mustafasrepo) [reply datafusion/physical-expr/src/aggregate/first_last.rs:506]: Indeed, however, here we are checking just a single row (row that have lowest value). Hence I don't think it is worth to conversion here.

File: datafusion/physical-expr/src/aggregate/first_last.rs (hunk 3 above)
Comment (alamb) [reply datafusion/physical-expr/src/aggregate/first_last.rs:539]: BTW I had the same basic need (find the position of min/max so I could find a value in a corresponding column) while implementing our [special `selector_first`, `selector_last`, etc functions in InfluxDB 3.0](https://docs.influxdata.com/influxdb/cloud-serverless/reference/sql/functions/selector/#selector-functions) (I also had to code them specially)

File: datafusion/physical-expr/src/aggregate/first_last.rs (hunk 6 above)
Comment (alamb) [reply datafusion/physical-expr/src/aggregate/first_last.rs:506]: I agree that since it is a single column max comparison this is probably fine (and no worse than the current implementation). If  we need to optimize performance we could probably implement specialized implementations (like `FirstValue<ArrowPrimitiveType>` and skip the copying entirely.

That is likely a premature optimization at this point

**Comment by alamb** [comment]
I plan to review this carefully either later today or tomorrow. I want to get a draft of https://github.com/apache/arrow-datafusion/issues/8491 first

File: datafusion/physical-expr/src/aggregate/first_last.rs (hunk 3 above)
Comment (ozankabak) [reply datafusion/physical-expr/src/aggregate/first_last.rs:539]: Do you think you implementation is more efficient? If that is the case, maybe we can use that code instead?

**Review by ozankabak — APPROVED** [review]
I reviewed this PR and it looks good to me. Looking forward to @alamb's thoughts

File: datafusion/physical-expr/src/aggregate/first_last.rs [hunk 7]
```diff
@@ -211,10 +212,45 @@ impl FirstValueAccumulator {
...
-        self.first = row[0].clone();
-        self.orderings = row[1..].to_vec();
-        self.is_set = true;
+    fn update_with_new_row(&mut self, row: &[ScalarValue]) -> Result<()> {
+        let [value, orderings @ ..] = row else {
+            return internal_err!("Empty row in FIRST_VALUE");
+        };
+        // Update when there is no entry in the state, or we have an "earlier"
+        // entry according to sort requirements.
+        if !self.is_set
```
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/first_last.rs:221]: In theory, we may be able to use a `Option<ScalarValue>` instead of `ScalarValue` and `is_set` flag, but I don't think it matters for performance and this PR follows the existing implementation as well 👍

File: datafusion/physical-expr/src/aggregate/first_last.rs (hunk 6 above)
Comment (alamb) [reply datafusion/physical-expr/src/aggregate/first_last.rs:504]: I re-reviewed and I agree that the `RowFormat` is not needed here (and in fact it may actually be slower) because, as @mustafasrepo  points out, this code uses `ScalarValue` to compare a single row per batch (it finds the largest/smallest row per batch using `lexsort_to_indices`). We would have to benchmark to be sure.

File: datafusion/physical-expr/src/aggregate/first_last.rs (hunk 3 above)
Comment (alamb) [reply datafusion/physical-expr/src/aggregate/first_last.rs:538]: I think our implementation is (slightly) more efficient, but it is less general (only works for timestamp columns). You can see the basic idea here

https://github.com/influxdata/influxdb/blob/main/query_functions/src/selectors.rs

And the comparision is here: https://github.com/influxdata/influxdb/blob/acfef87659c9a8c4c49e4628264369569e04cad1/query_functions/src/selectors/internal.rs#L119-L127

I think we should stay with the `ScalarValue` implementation unless we find some query where this calculation is taking most of the time

File: datafusion/physical-expr/src/aggregate/mod.rs [hunk 8]
```diff
@@ -134,10 +137,7 @@ pub trait AggregateExpr: Send + Sync + Debug + PartialEq<dyn Any> {
...
 /// Checks whether the given aggregate expression is order-sensitive.
 /// For instance, a `SUM` aggregation doesn't depend on the order of its inputs.
-/// However, a `FirstValue` depends on the input ordering (if the order changes,
-/// the first value in the list would change).
+/// However, an `ARRAY_AGG` with `ORDER BY` depends on the input ordering.
 pub fn is_order_sensitive(aggr_expr: &Arc<dyn AggregateExpr>) -> bool {
-    aggr_expr.as_any().is::<FirstValue>()
-        || aggr_expr.as_any().is::<LastValue>()
-        || aggr_expr.as_any().is::<OrderSensitiveArrayAgg>()
+    aggr_expr.as_any().is::<OrderSensitiveArrayAgg>()
```
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/mod.rs:142]: Eventually this would be a nice thing to move into the `AggregateExpr` trait directly so we could override it and avoid special casing built in functions. Not for this PR though :)

File: datafusion/sqllogictest/test_files/distinct_on.slt [hunk 9]
```diff
@@ -100,10 +100,9 @@ ProjectionExec: expr=[FIRST_VALUE(aggregate_test_100.c3) ORDER BY [aggregate_tes
 ------AggregateExec: mode=FinalPartitioned, gby=[c1@0 as c1], aggr=[FIRST_VALUE(aggregate_test_100.c3), FIRST_VALUE(aggregate_test_100.c2)]
 --------CoalesceBatchesExec: target_batch_size=8192
 ----------RepartitionExec: partitioning=Hash([c1@0], 4), input_partitions=4
-------------AggregateExec: mode=Partial, gby=[c1@0 as c1], aggr=[FIRST_VALUE(aggregate_test_100.c3), FIRST_VALUE(aggregate_test_100.c2)], ordering_mode=Sorted
---------------SortExec: expr=[c1@0 ASC NULLS LAST,c3@2 ASC NULLS LAST]
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/distinct_on.slt:104]: I do love the lack of Sort here

**Review by alamb — APPROVED** [review]
Thank you @mustafasrepo and @ozankabak   -- this PR looks good to me. ❤️ 

I believe if we applied the same change to `ArrayAgg` I think we can remove the limitation of a single compatible `ORDER BY` in a query -- aka https://github.com/apache/arrow-datafusion/issues/8582 -- is that your understanding too?

I am sorry for the delay in reviewing, I am partly on holiday this week so don't have as much time to devote to these endeavors as normal. 

I think that in many common queries, this implementation is likely faster than what is on main because it doesn't potentially re-sort the entire input (it instead used `lexsort_to_indices`)

As we discussed in the design document the potential downside of this approach is that if multiple aggregates share the same `ORDER BY` clause, they will each independently sort the input batches, which is unfortunate but could be optimized in future PTs

File: datafusion/physical-expr/src/aggregate/mod.rs (hunk 8 above)
Comment (ozankabak) [reply datafusion/physical-expr/src/aggregate/mod.rs:142]: I like this idea 👍

**Comment by ozankabak** [comment]
> I believe if we applied the same change to ArrayAgg I think we can remove the limitation of a single compatible ORDER BY in a query -- aka https://github.com/apache/arrow-datafusion/issues/8582 -- is that your understanding too?

Yes, if we make `ARRAY_AGG` sort internally, we can do this

> As we discussed in the design document the potential downside of this approach is that if multiple aggregates share the same ORDER BY clause, they will each independently sort the input batches, which is unfortunate but could be optimized in future PTs

Exactly. We will explore both the split/diamond approach and the approach above in the upcoming weeks
